content/blog/next-level-iac-breakpoint-debugging/index.md

```diff
@@ -0,0 +1,317 @@
+
+```json
+{
+	"version": "2.0.0",
+	"tasks": [
+	    {
+			"label": "pulumi-debug",
+			"type": "shell",
+			"command": ". ./pulumi-debug.sh",
+			"isBackground": true,
+            "options": {
+                "cwd": "${workspaceFolder}"
+            },
+
+		}
+	]
+}
+```
+
+This creates a VS Code task named `pulumi-debug` which will run the wrapper script we just created from the root of our project.
+
+Next, create a file named `launch.json` in the same directory. This will configure the behavior of the `F5` key!
+
+```json
+{
+    "version": "0.2.0",
+    "configurations": [
+        {
+            "name": "Launch Pulumi Program (debug)",
+            "type": "node",
+            "port": 9292,
+            "request": "attach",
+            "preLaunchTask": "pulumi-debug",
+            "continueOnAttach": true,
+            "restart": {
+                "delay": 5000,
+                "maxAttempts": 5
+            },
+            "skipFiles": [
+                "${workspaceFolder}/node_modules/**/*.js",
+                "${workspaceFolder}/lib/**/*.js",
+                "<node_internals>/**/*.js"
+            ]
+        }
+    ]
+}
+```
+{{% notes type="info" %}}
+Let's break this down a bit: 
+
+* `type: node` tells VS Code that we are using the Node runtime
+* `port: 9292` tells VS Code to attach the debugger to port 9292
+* `request: attach` tells VS Code that we want to attach to to the debugger at launch
+* `preLaunchTask: pulumi-debug` tells VS Code to run the task we configured in `tasks.json` before we connect the debugger. This will run our `pulumi-debug.sh` script and then attach to the Node.js runtime 
+* `continueOnAttach: true` tells VS Code to start running our code as soon as we attach to the waiting Node.js runtime (until we reach our first breakpoint that is, where it will pause and return control to the debugger client, aka VS Code)
+* `restart: {...}` configures VS Code to wait 5 seconds and then retry connecting, up to 5 times. This is here in case it takes a while for the Node.js runtime to start running. It's unlikely, but possible!
+* `skipFiles: {...}` sets up some filters of files *not* to try to debug. This includes everything that is part of node's internals, 3rdparty node modules, and other libraries. That way we don't end up deep in someone else's code unexpectedly and can just focus on our Pulumi program.
+* `name` sets a human-readable name which shows up in the drop down box in VS Code when you choose which [launch configuration][vs-code-launch-configs] to run. 
+{{% /notes %}}
+
+Finally, we need to make one more edit to the `pulumi-debug.sh` file we created. Here will will pass the necessary options to the Node.js runtime to set it to run in debug mode. 
+
+Add the following line to the beginning of the file:
+
+```shell {hl_lines=[3]}
+#!/bin/bash
+
+export NODE_OPTIONS='--inspect-brk=127.0.0.1:9292'
```

Comment:
i believe we need to do a python version of this since all our blog posts should have both typescript and python code

Reply:
In this case, the blog post is already so complex, and the differences between Python and Node/TS here are significant enough that I think it warrants a separate follow up post. Similarly w/ Go and C#...